[SPARK-20758][SQL] Add Constant propagation optimization

tejasapatil opened the pull request:
## What changes were proposed in this pull request?

See class doc of `ConstantPropagation` for the approach used.

## How was this patch tested?

- Added unit tests

tejasapatil committed:
Initial version of constant propagation

tejasapatil commented:
Jenkins test this please

hvanhovell commented:
ok to test

hvanhovell commented:
it is weird that jenkins is not kicking off

SparkQA commented:
**[Test build #76964 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76964/testReport)** for PR 17993 at commit [`bb3b349`](https://github.com/apache/spark/commit/bb3b349d5b895bd8ca5b7db75845dede08601f40).

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:
```diff
@@ -54,6 +54,59 @@ object ConstantFolding extends Rule[LogicalPlan] {
+ */
+object ConstantPropagation extends Rule[LogicalPlan] with PredicateHelper {
+
+  def containsNonConjunctionPredicates(expression: Expression): Boolean = expression match {
```
This might be more straightfoward:
```scala
expression.find {
  case _: Not | _: Or => true
}.isDefined
```

tejasapatil replied:
did this change

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:
```diff
@@ -54,6 +54,59 @@ object ConstantFolding extends Rule[LogicalPlan] {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case q: LogicalPlan => q transformExpressionsUp {
+      case and @ (left And right)
```
`case and: And if containsNonConjunctionPredicates(and)`?

tejasapatil replied:
did this change

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:
```diff
@@ -54,6 +54,59 @@ object ConstantFolding extends Rule[LogicalPlan] {
+      case and @ (left And right)
+        if !containsNonConjunctionPredicates(left) && !containsNonConjunctionPredicates(right) =>
+
+        val leftEntries = left.collect {
```
Lets put the collect in a function, so we can avoid the repetition.

tejasapatil replied:
sure

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:
```diff
@@ -54,6 +54,59 @@ object ConstantFolding extends Rule[LogicalPlan] {
+        val predicates = (leftEntries.map(_._2) ++ rightEntries.map(_._2)).toSet
+
+        def replaceConstants(expression: Expression) = expression transform {
+          case a: AttributeReference if constantsMap.contains(a) =>
```
I don't think the double lookup is necessary. `constantsMap.get(a).getOrElse(a)` should cover this.

hvanhovell replied:
What happens if I do something stupid like `i = 1 and ((j = 1) = (j = i))`? I think `j = 1` might replaced by `1 = 1`.

tejasapatil replied:
Nice catch !!! I changed the logic to handle that.

SparkQA commented:
**[Test build #76964 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76964/testReport)** for PR 17993 at commit [`bb3b349`](https://github.com/apache/spark/commit/bb3b349d5b895bd8ca5b7db75845dede08601f40).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76964/
Test FAILed.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:
```diff
@@ -54,6 +54,59 @@ object ConstantFolding extends Rule[LogicalPlan] {
+
+        and transform {
+          case e @ EqualTo(_, _) if !predicates.contains(e) &&
+            e.references.exists(ref => constantsMap.contains(ref)) =>
```
Building the `references` map is more expensive, shall we just skip this?

tejasapatil replied:
skipped it

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:
```diff
@@ -54,6 +54,59 @@ object ConstantFolding extends Rule[LogicalPlan] {
+        if !containsNonConjunctionPredicates(left) && !containsNonConjunctionPredicates(right) =>
+
+        val leftEntries = left.collect {
+          case e @ EqualTo(left: AttributeReference, right: Literal) => ((left, right), e)
```
How about `EqualNullSafe`? Normally, we use `Equality`

tejasapatil replied:
did this change

tejasapatil committed:
review comment

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ConstantPropagationSuite.scala:
```diff
@@ -0,0 +1,102 @@
+  private val columnB = 'b.int
+
+  /**
+   * Unit tests for constant propagation in expressions.
```
Hi, @tejasapatil . Nit. It looks like test suite comment. Can we move this comment to line 27?

tejasapatil replied:
did this change

SparkQA commented:
**[Test build #77133 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77133/testReport)** for PR 17993 at commit [`cc026da`](https://github.com/apache/spark/commit/cc026da840714bc2f88076dbb2aafa70aa1fa0b7).

SparkQA commented:
**[Test build #77134 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77134/testReport)** for PR 17993 at commit [`b8c4147`](https://github.com/apache/spark/commit/b8c4147d3b7dd2c1d0e6b3015042271e754a18cf).

tejasapatil commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:
```diff
@@ -54,6 +54,62 @@ object ConstantFolding extends Rule[LogicalPlan] {
+  }.isDefined
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case f: Filter => f transformExpressionsUp {
```
I was initially doing this for the entire logical plan but now switched to do only for filter operator. 
Reason: Doing this for the entire logical plan will mess up with JOIN predicates. eg.

```
SELECT * FROM a JOIN b ON a.i = 1 AND b.i = a.i
=>
 SELECT * FROM a JOIN b ON a.i = 1 AND b.i = 1
```

.. the result is a cartesian product and Spark fails (asking to set a config). In case of OUTER JOINs, changing the join predicates might cause regression.

hvanhovell replied:
Maybe I am being myopic here but the result should be the same right? The only way this regresses is when we plan a `CartesianProduct` instead of an `BroadcastNestedLoopJoin`... I am fine with not optimizing this for now, it would be nice if these constraints are at least generated here.

tejasapatil replied:
Yes the result should be the same. I don't have any theoretical proof if doing this over joins will be safe so want to be cautious here ... any bad rules might lead to correctness bugs which is super bad for end users.

> it would be nice if these constraints are at least generated here

Sorry I am not able to get you here and want to make sure if I am not ignoring your comment. Are you suggesting any changes over the existing version ?

hvanhovell replied:
We currently infer `is not null` constraints up and down the plan. This could be easily extended to other constraints. Your PR has some overlap with this. However, lets focus on getting this merged first, and then we might take a stab at extending this.

hvanhovell replied:
also cc @sameeragarwal

SparkQA commented:
**[Test build #77133 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77133/testReport)** for PR 17993 at commit [`cc026da`](https://github.com/apache/spark/commit/cc026da840714bc2f88076dbb2aafa70aa1fa0b7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77133/
Test FAILed.

SparkQA commented:
**[Test build #77134 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77134/testReport)** for PR 17993 at commit [`b8c4147`](https://github.com/apache/spark/commit/b8c4147d3b7dd2c1d0e6b3015042271e754a18cf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77134/
Test FAILed.

tejasapatil committed:
Fix unit tests

- FileSourceStrategySuite.partitioned table
- FileSourceStrategySuite.partitioned table - case insensitive
- FileSourceStrategySuite.partitioned table - after scan filters

tejasapatil commented:
Jenkins test this please

SparkQA commented:
**[Test build #77160 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77160/testReport)** for PR 17993 at commit [`aaad78c`](https://github.com/apache/spark/commit/aaad78c4ab0985812798ea9b19fff5c9a4808fdc).

tejasapatil commented:
Jenkins test this please

SparkQA commented:
**[Test build #77193 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77193/testReport)** for PR 17993 at commit [`aaad78c`](https://github.com/apache/spark/commit/aaad78c4ab0985812798ea9b19fff5c9a4808fdc).

SparkQA commented:
**[Test build #77193 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77193/testReport)** for PR 17993 at commit [`aaad78c`](https://github.com/apache/spark/commit/aaad78c4ab0985812798ea9b19fff5c9a4808fdc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77193/
Test FAILed.

tejasapatil committed:
unit test

SparkQA commented:
**[Test build #77392 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77392/testReport)** for PR 17993 at commit [`399f348`](https://github.com/apache/spark/commit/399f34859e57b7330e39bccdc33b9cd602227bda).

SparkQA commented:
**[Test build #77392 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77392/testReport)** for PR 17993 at commit [`399f348`](https://github.com/apache/spark/commit/399f34859e57b7330e39bccdc33b9cd602227bda).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77392/
Test PASSed.

tejasapatil commented:
cc @hvanhovell @gatorsmile @dongjoon-hyun

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:
```diff
@@ -54,6 +54,62 @@ object ConstantFolding extends Rule[LogicalPlan] {
+ *   in the AND node.
+ */
+object ConstantPropagation extends Rule[LogicalPlan] with PredicateHelper {
+  def containsNonConjunctionPredicates(expression: Expression): Boolean = expression.find {
```
`private def`?

tejasapatil replied:
done

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:
```diff
@@ -54,6 +54,62 @@ object ConstantFolding extends Rule[LogicalPlan] {
+        }
+
+        val constantsMap = AttributeMap(equalityPredicates.map(_._1))
+        val predicates = equalityPredicates.map(_._2).toSet
```
I'm wondering if it's safe when we have both `a = 1` and `a = 2` at the same time?

tejasapatil replied:
Current impl will pick the last one (ie. `a = 2`) and propagate it. Given that its one of the equality predicates user provided, there is nothing wrong in propagating it. When the query is evaluated, it would return empty result given that `a = 1` and `a = 2` cannot be true at the same time.

```
scala> hc.sql(" SELECT * FROM table1 a WHERE a.j = 1 AND a.j = 2 AND a.k = (a.j + 3)").explain(true)

== Physical Plan ==
*Project [i#51, j#52, k#53]
+- *Filter ((((isnotnull(k#53) && isnotnull(j#52)) && (j#52 = 1)) && (j#52 = 2)) && (cast(k#53 as int) = 5))
   +- *FileScan orc default.table1[i#51,j#52,k#53] Batched: false, Format: ORC, Location: InMemoryFileIndex[file:/Users/tejasp/warehouse/table1], PartitionFilters: [], PushedFilters: [IsNotNull(k), IsNotNull(j), EqualTo(j,1), EqualTo(j,2)], ReadSchema: struct<i:int,j:int,k:string>
```

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ConstantPropagationSuite.scala:
```diff
@@ -0,0 +1,154 @@
+        .where(columnA === Literal(11) && columnB === Literal(10)).analyze
+
+    comparePlans(Optimize.execute(query.analyze), correctAnswer)
+  }
```
Could you add a negative test case like `SELECT * FROM t WHERE a=1 and a=2 and b=a+3`?

tejasapatil replied:
added

tejasapatil committed:
more test case

SparkQA commented:
**[Test build #77432 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77432/testReport)** for PR 17993 at commit [`731f796`](https://github.com/apache/spark/commit/731f79626b38321845ccca8813174a9cfa2331dd).

SparkQA commented:
**[Test build #77432 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77432/testReport)** for PR 17993 at commit [`731f796`](https://github.com/apache/spark/commit/731f79626b38321845ccca8813174a9cfa2331dd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77432/
Test PASSed.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:
```diff
@@ -54,6 +54,62 @@ object ConstantFolding extends Rule[LogicalPlan] {
+        }
+
+        and transform {
+          case e @ EqualTo(_, _) if !predicates.contains(e) => replaceConstants(e)
```
Should we check for identity instead of equality? I think you are doing the latter. What will happen in the following example: `select * from bla where (a = 1 or b = 2) and a = 1`

tejasapatil replied:
Here is the behavior with the PR. Seems reasonable because `a = 1` has to be true so `(a = 1 or b = 2)` would always be `true` and can be eliminated.

```
scala> hc.sql(" select * from bla where (a = 1 or b = 2) and a = 1 ").explain(true)

== Physical Plan ==
*Project [a#34, b#35]
+- *Filter (isnotnull(a#34) && (a#34 = 1))
   +- *FileScan ....
```

hvanhovell commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ConstantPropagationSuite.scala:
```diff
@@ -0,0 +1,167 @@
+        .where(
+          columnA === Literal(11) &&
+            columnB === Literal(10) &&
+            (columnA === Add(columnC, Literal(3)) || Literal(10) === columnC))
```
Should we be able to infer that `columnA == Literal(11)`?

hvanhovell replied:
Perhaps if you increase the number of iterations on `ConstantPropagation` batch...

tejasapatil replied:
I had set a higher value for iterations in previous version of this PR but somehow the unit tests kept failing for me over terminal (surprisingly they worked fine over Intellij). This seems unrelated to the change done in the PR. If you have any advice here, let me know

eg. https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77193/testReport/org.apache.spark.sql.catalyst.optimizer/ConstantPropagationSuite/basic_test/ 

```
sbt.ForkMain$ForkError: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: Max iterations (2) reached for batch ConstantPropagation, tree:
!Filter ((a#82440 = 11) && (b#82441 = 10))
+- !Project [a#82440]
   +- LocalRelation <empty>, [a#82437, b#82438, c#82439]

	at org.apache.spark.sql.catalyst.rules.RuleExecutor$$anonfun$execute$1.apply(RuleExecutor.scala:105)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor$$anonfun$execute$1.apply(RuleExecutor.scala:74)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.execute(RuleExecutor.scala:74)
	at org.apache.spark.sql.catalyst.optimizer.ConstantPropagationSuite$$anonfun$1.apply$mcV$sp(ConstantPropagationSuite.scala:60)
	at org.apache.spark.sql.catalyst.optimizer.ConstantPropagationSuite$$anonfun$1.apply(ConstantPropagationSuite.scala:50)
```

hvanhovell replied:
Hmmm... that means the optimizer is not converging to a fixed point. Could you try to increase the number of iterations? You can also check if the optimizer reaches 100 iterations during regular execution; it should log a warning. If it does something is wrong with the rule, and it might cause the optimizer to run prohibitively long...

tejasapatil replied:
That did it !! updated the change

tejasapatil committed:
test case: change num iterations

SparkQA commented:
**[Test build #77493 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77493/testReport)** for PR 17993 at commit [`38543de`](https://github.com/apache/spark/commit/38543de8b45dc4145182d20e84a5796579548f3f).

SparkQA commented:
**[Test build #77493 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/77493/testReport)** for PR 17993 at commit [`38543de`](https://github.com/apache/spark/commit/38543de8b45dc4145182d20e84a5796579548f3f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/77493/
Test PASSed.

hvanhovell commented:
LGTM - merging to master. Thanks!
